File: java/src/strategy/SimUDuckApp/Revision3/MiniDuckSimulator.java
package strategy.SimUDuckApp.Revision3;

import java.util.ArrayList;

public class MiniDuckSimulator {

	public static void main(String[] args) {
		
		ArrayList<Duck> ducks = new ArrayList<Duck>();
		
		ducks.add(new MallardDuck());
		ducks.add(new RedheadDuck());
		ducks.add(new RubberDuck());
		ducks.add(new DecoyDuck());
		
		for (Duck d: ducks) {
			d.display();
			d.swim();
			if (d instanceof Quackable)
				((Quackable)d).quack();
			if (d instanceof Flyable)
				((Flyable)d).fly();
			System.out.println();
		}

	}

}

File: src/strategy/SimUDuckApp/Revision3/MiniDuckSimulator.kt
package strategy.SimUDuckApp.Revision3


fun main() {
    val ducks = ArrayList<Duck>()
    val d: MallardDuck = MallardDuck()

    ducks.add(MallardDuck())
    ducks.add(RedheadDuck())
    ducks.add(RubberDuck())
    ducks.add(DecoyDuck())

    for (d in ducks) {
        d.display()
        d.swim()
        if (d is Quackable) (d as Quackable).quack()
        if (d is Flyable) (d as Flyable).fly()
        println()
    }
}

